consensus/istanbul/backend/announce.go

```diff
@@ -155,9 +213,80 @@ func (sb *Backend) sendIstAnnounce() error {
 	enodeUrl := enode.String()
 	view := sb.core.CurrentView()
 
-	msg := &announceMessage{Address: sb.Address(),
-		EnodeURL: enodeUrl,
-		View:     view}
+	regVals, err := sb.retrieveRegisteredValidators()
+	// The validator contract may not be deployed yet.
+	// Even if it is deployed, it may not have any registered validators yet.
+	if err == errValidatorsContractNotRegistered || len(regVals) == 0 {
+		sb.logger.Trace("Can't retrieve the registered validators.  Only allowing the initial validator set to send announce messages", "err", err, "regVals", regVals)
+		valSet := sb.getValidators(block.Number().Uint64(), block.Hash())
+
+		regVals = make(map[common.Address]bool)
+		for _, val := range valSet.List() {
+			regVals[val.Address()] = true
+		}
+	} else if err != nil {
+		sb.logger.Error("Error in retrieving the registered validators", "err", err)
+		return err
+	}
+
+	useValidatorContract := true
+	var g core.Genesis
+	validatorsAddress, err := sb.regAdd.GetRegisteredAddressAtCurrentHeader(params.ValidatorsRegistryId)
+	if err != nil {
+		log.Warn("Registry address lookup failed", "err", err)
+		useValidatorContract = false
+		genesisJSON, err := sb.db.Get(core.DBGenesisKey)
+		err = g.UnmarshalJSON(genesisJSON)
+		if err != nil {
+			log.Error("Unable to retrieve genesis from db", "err", err)
+		}
+	}
+
+	encryptedEnodeUrls := make(map[common.Address][]byte)
+	for addr := range regVals {
+		var pubKeyBytes []byte
+		if useValidatorContract {
+			state, err := sb.stateAt(block.Header().ParentHash)
+			if err != nil {
+				log.Error("verify - Error in getting the block's parent's state", "parentHash", block.Header().ParentHash.Hex(), "err", err)
+				return err
+			}
+			var validator []interface{}
+			if _, err := sb.iEvmH.MakeStaticCall(*validatorsAddress, getValidatorFuncABI, "getValidator", []interface{}{}, &validator, uint64(1000000), block.Header(), state); err != nil {
+				log.Error("Unable to retrieve Validator Account from Validator smart contract", "err", err)
+				return err
+			}
+			pubKeyBytes, _ = validator[3].([]byte) // The publicKey field
+		} else {
+			pubKeyBytes = g.Alloc[addr].PublicKey
+		}
+		// ECDSAKey, err := crypto.UnmarshalPubkey(pubKeyBytes)
+		// if err != nil {
+		// 	log.Info("bigsad")
+		// 	continue;
+		// }
+		// pubKey := ecies.ImportECDSAPublic(ECDSAKey)
+		pubKey, err := p2p.ImportPublicKey(pubKeyBytes)
+		encryptedEnodeUrl, err := ecies.Encrypt(rand.Reader, pubKey, []byte(enodeUrl), nil, nil)
+		if err != nil {
+			log.Error("Unable to unmarshal public key", "err", err)
+		} else {
+			encryptedEnodeUrls[addr] = encryptedEnodeUrl
+		}
+	}
+
+	encryptedEnodeData, err := json.Marshal(encryptedEnodeUrls)
```

Comment:
Do we need to serialize the `encryptedEnodeUrls`?  The whole announceMessage is going to be serialized.

Reply:
I serialized it using `json` because the serialization we do for the object does not serialize go maps

Reply:
Seems like the ethereum wiki does have a suggestion on how to encode maps:  https://github.com/ethereum/wiki/wiki/Design-Rationale#rlp.

So EncodeRLP could transform the go map as described in the wiki, and DecodeRLP would then do the opposite transformation.

In general, all of the networking messages used within Ethereum uses RLP encoding, so for consistency, it would be nice if this field does as well.